File: languages/csharp/04-expert/smart-pointers-in-depth/example/main.cs
// Module focus: Tracking ownership and lifetime when multiple references can observe the same value.
// Why it matters: practicing smart pointers in depth patterns makes exercises and checkpoints easier to reason about.

using System;

// Helper setup for smart pointers in depth; this keeps the walkthrough readable.
sealed record Report(string Title);

sealed class ReportOwner
{
    public ReportOwner(string name, Report? current) => (Name, Current) = (name, current);

    public string Name { get; }
    public Report? Current { get; private set; }

    public void TransferTo(ReportOwner destination)
    {
        // Empty owners cannot transfer ownership, so make that state explicit.
        if (Current is null)
        {
            Console.WriteLine($"{Name} has nothing to transfer.");
            return;
        }

        // Report output values so learners can verify the smart pointers in depth result.
        Console.WriteLine($"{Name} transfers {Current.Title} to {destination.Name}.");
        destination.Current = Current;
        Current = null;
    }

    public void Describe() =>
        Console.WriteLine($"{Name}: {(Current is null ? "empty" : Current.Title)}");
}

sealed class PreviewPane
{
    private readonly WeakReference<Report> currentReport;

    public PreviewPane(Report report) => currentReport = new WeakReference<Report>(report);

    public void Describe()
    {
        // Weak observers must check whether the target is still alive.
        if (currentReport.TryGetTarget(out Report? report))
        {
            Console.WriteLine($"Preview can still see: {report.Title}");
            return;
        }

        Console.WriteLine("Preview target expired.");
    }
}

class Program
{
    // Walk through ownership transfer and weak observation in one fixed scenario.
    static void Main()
    {
        // Start with one owner holding the report and one empty destination.
        ReportOwner inbox = new ReportOwner("Inbox", new Report("Quarterly Summary"));
        ReportOwner archive = new ReportOwner("Archive", null);

        // After transfer, the source is empty and the destination owns the report.
        inbox.Describe();
        inbox.TransferTo(archive);
        inbox.Describe();
        archive.Describe();

        PreviewPane preview = CreatePreview();
        preview.Describe();

        // Force collection so the weak-reference branch is visible in this tiny example.
        GC.Collect();
        GC.WaitForPendingFinalizers();
        GC.Collect();

        preview.Describe();
    }

    static PreviewPane CreatePreview()
    {
        // The preview does not keep this transient report alive after the method returns.
        return new PreviewPane(new Report("Transient Draft"));
    }
}

File: languages/java/04-expert/smart-pointers-in-depth/example/Main.java
// Module focus: Tracking ownership and lifetime when multiple references can observe the same value.
// Why it matters: the example makes it possible to model exclusive, shared, and non-owning
// relationships idiomatically before the learner tackles the exercises.

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Main {
    record Report(String title) {
    }

    static final class ReportOwner {
        private final String name;
        private Report current;

        ReportOwner(String name, Report current) {
            this.name = name;
            this.current = current;
        }

        Optional<Report> current() {
            // Optional keeps "empty owner" checks explicit at API boundaries.
            return Optional.ofNullable(current);
        }

        void transferTo(ReportOwner destination) {
            // Empty owners cannot transfer ownership, so the method exits visibly.
            if (current == null) {
                System.out.println(name + " has nothing to transfer.");
                return;
            }

            // Moving the reference leaves the source empty and the destination responsible.
            System.out.println(name + " transfers " + current.title() + " to " + destination.name + ".");
            destination.current = current;
            current = null;
        }

        void describe() {
            System.out.println(name + ": " + current().map(Report::title).orElse("empty"));
        }
    }

    static final class ReportArchive {
        private final List<Report> reports = new ArrayList<>();

        void add(Report report) {
            // Strong references in this list keep archived reports alive.
            reports.add(report);
        }

        List<Report> snapshot() {
            // A defensive snapshot prevents callers from mutating archive internals.
            return List.copyOf(reports);
        }
    }

    public static void main(String[] args) {
        ReportOwner inbox = new ReportOwner("Inbox", new Report("Quarterly Summary"));
        ReportOwner archiveSlot = new ReportOwner("Archive", null);

        // Report ownership before and after transfer so the lifetime change is visible.
        inbox.describe();
        archiveSlot.describe();
        inbox.transferTo(archiveSlot);
        inbox.describe();
        archiveSlot.describe();

        ReportArchive archive = new ReportArchive();
        archiveSlot.current().ifPresent(archive::add);
        archive.add(new Report("Budget Notes"));
        // The snapshot is safe to share because callers cannot mutate archive internals.
        System.out.println("Archive snapshot size: " + archive.snapshot().size());

    }
}
